File: sdk/android/frak-sdk/src/main/kotlin/id/frak/sdk/identity/ProofCodec.kt
package id.frak.sdk.identity

import id.frak.sdk.core.Base64Url
import id.frak.sdk.core.Uuid

/** Domain-separated proof operations: a signature for one op must never verify for another. */
internal enum class ProofOp(
    val wireValue: String,
) {
    Ensure("frak-ensure-v1"),
    Install("frak-install-v1"),
    Merge("frak-merge-v1"),
}

/**
 * Frozen wire format for identity proof-of-possession, fixed-width byte copies (no JSON):
 * ```text
 * msg      := op ‖ merchantId(16) ‖ anonymousId(16) ‖ binding(32) ‖ ts(8)
 * envelope := v(1) ‖ pk(65) ‖ ts(8) ‖ sig(64)
 * proof    := base64url(envelope), unpadded
 * ```
 * UUIDs signed as raw 16 bytes, never the 36-char text form.
 */
internal object ProofCodec {
    const val ENVELOPE_VERSION: Int = 1

    const val UUID_BYTES: Int = 16
    const val BINDING_BYTES: Int = 32
    const val TS_BYTES: Int = 8

    /** `0x04` prefix plus two 32-byte coordinates. */
    const val PUBKEY_BYTES: Int = 65

    /** Raw `r‖s` ECDSA; low-S normalisation not guaranteed. */
    const val SIG_BYTES: Int = 64

    /** Parses (not lowercases) so signed bytes never depend on the caller's case formatting. */
    fun uuidToBytes(
        value: String,
        label: String,
    ): ByteArray = Uuid.toBytes(value, label)

    /** Formats 16 raw bytes as a lowercase hyphenated UUID. */
    fun bytesToUuid(
        bytes: ByteArray,
        offset: Int = 0,
    ): String = Uuid.fromBytes(bytes, offset)

    /** First 16 bytes of the SHA-256 digest, RFC-4122 version/variant bits overwritten. */
    fun deriveClientIdFromHash(hash: ByteArray): String {
        require(hash.size >= UUID_BYTES) {
            "deriveClientIdFromHash requires at least $UUID_BYTES bytes, got ${hash.size}"
        }
        val bytes = hash.copyOf(UUID_BYTES)
        bytes[6] = ((bytes[6].toInt() and 0x0F) or 0x40).toByte()
        bytes[8] = ((bytes[8].toInt() and 0x3F) or 0x80).toByte()
        return bytesToUuid(bytes)
    }

    /** [binding] must be empty (written as 32 zero bytes) or exactly [BINDING_BYTES]. */
    fun buildMessage(
        op: ProofOp,
        merchantId: String,
        anonymousId: String,
        binding: ByteArray,
        ts: Long,
    ): ByteArray {
        require(binding.isEmpty() || binding.size == BINDING_BYTES) {
            "binding must be empty or $BINDING_BYTES bytes, got ${binding.size}"
        }

        val opBytes = op.wireValue.toByteArray(Charsets.US_ASCII)
        val out = ByteArray(opBytes.size + UUID_BYTES * 2 + BINDING_BYTES + TS_BYTES)

        var offset = 0
        opBytes.copyInto(out, offset)
        offset += opBytes.size
        uuidToBytes(merchantId, "merchantId").copyInto(out, offset)
        offset += UUID_BYTES
        uuidToBytes(anonymousId, "anonymousId").copyInto(out, offset)
        offset += UUID_BYTES
        binding.copyInto(out, offset)
        offset += BINDING_BYTES
        writeUint64Be(out, offset, ts)

        return out
    }

    fun encodeProof(
        publicKey: ByteArray,
        ts: Long,
        signature: ByteArray,
    ): String {
        require(publicKey.size == PUBKEY_BYTES) { "pk must be $PUBKEY_BYTES bytes, got ${publicKey.size}" }
        require(signature.size == SIG_BYTES) { "sig must be $SIG_BYTES bytes, got ${signature.size}" }

        val out = ByteArray(1 + PUBKEY_BYTES + TS_BYTES + SIG_BYTES)
        out[0] = ENVELOPE_VERSION.toByte()
        publicKey.copyInto(out, 1)
        writeUint64Be(out, 1 + PUBKEY_BYTES, ts)
        signature.copyInto(out, 1 + PUBKEY_BYTES + TS_BYTES)
        return Base64Url.encode(out)
    }

    private fun writeUint64Be(
        target: ByteArray,
        offset: Int,
        value: Long,
    ) {
        require(value >= 0) { "ts must be a non-negative integer: $value" }
        for (index in 0 until TS_BYTES) {
            target[offset + index] = (value ushr ((TS_BYTES - 1 - index) * 8) and 0xFF).toByte()
        }
    }
}

File: sdk/ios/Sources/FrakSDK/Identity/ProofCodec.swift
import Foundation

// Raw value is the domain separator prepended to signed bytes. Only what native can mint:
// `frak-ensure-v1` needs a wallet session token no native client holds (the install deep link is
// the equivalent) and there is no native SSO surface.
enum ProofOp: String {
    case install = "frak-install-v1"
    case merge = "frak-merge-v1"
}

// Never merchant-facing: exists so the reason survives to the log line.
struct InvalidProofInput: Error, CustomStringConvertible {
    let description: String
}

// Identity proof wire format, pinned to sdk/core/src/identity/fixtures/golden-proofs.json.
// Both layouts fixed-width, no length prefixes:
// message  := op_ascii + merchantId(16) + anonymousId(16) + binding(32) + ts_uint64be(8)
// envelope := version(1) + pubkey(65) + ts_uint64be(8) + signature(64)
// UUIDs signed as raw 16 bytes, never the 36-char text form.
enum ProofCodec {
    static let envelopeVersion: UInt8 = 1
    // Uncompressed SEC 1 point: 0x04 + X(32) + Y(32).
    static let publicKeyBytes = 65
    // Raw r+s, not DER.
    static let signatureBytes = 64
    private static let uuidByteCount = 16
    private static let bindingBytes = 32

    static func uuidBytes(_ value: String, label: String) throws -> Data {
        guard let uuid = UUID(uuidString: value) else {
            throw InvalidProofInput(description: "\(label) must be a UUID string, got: \(value)")
        }
        return withUnsafeBytes(of: uuid.uuid) { Data($0) }
    }

    // Lowercase always: UUID.uuidString is uppercase, and case mismatch as a string
    // breaks cache keys and the self-referral guard.
    static func uuidString(_ bytes: Data) -> String {
        let digits = Array(Hex.encode(bytes))
        return [0..<8, 8..<12, 12..<16, 16..<20, 20..<32]
            .map { String(digits[$0]) }
            .joined(separator: "-")
    }

    // Derives anonymous id from SHA-256(pubkey), with RFC-4122 v4 bits set.
    static func clientId(fromHash hash: Data) throws -> String {
        guard hash.count >= uuidByteCount else {
            throw InvalidProofInput(description: "clientId derivation needs at least 16 bytes, got \(hash.count)")
        }
        var bytes = [UInt8](hash.prefix(uuidByteCount))
        bytes[6] = bytes[6] & 0x0F | 0x40
        bytes[8] = bytes[8] & 0x3F | 0x80
        return uuidString(Data(bytes))
    }

    // binding is either empty (written as 32 zero bytes) or exactly 32.
    static func message(
        op: ProofOp,
        merchantId: String,
        anonymousId: String,
        binding: Data,
        ts: Int64
    ) throws -> Data {
        guard binding.isEmpty || binding.count == bindingBytes else {
            throw InvalidProofInput(description: "binding must be empty or 32 bytes, got \(binding.count)")
        }
        var out = Data(op.rawValue.utf8)
        out += try uuidBytes(merchantId, label: "merchantId")
        out += try uuidBytes(anonymousId, label: "anonymousId")
        out += binding.isEmpty ? Data(repeating: 0, count: bindingBytes) : binding
        out += try bigEndian(ts)
        return out
    }

    static func proof(publicKey: Data, ts: Int64, signature: Data) throws -> String {
        guard publicKey.count == publicKeyBytes else {
            throw InvalidProofInput(description: "pk must be 65 bytes, got \(publicKey.count)")
        }
        guard signature.count == signatureBytes else {
            throw InvalidProofInput(description: "sig must be 64 bytes, got \(signature.count)")
        }
        var out = Data([envelopeVersion])
        out += publicKey
        out += try bigEndian(ts)
        out += signature
        return Base64URL.encode(out)
    }

    private static func bigEndian(_ value: Int64) throws -> Data {
        guard value >= 0 else {
            throw InvalidProofInput(description: "ts must be a non-negative integer: \(value)")
        }
        return withUnsafeBytes(of: UInt64(value).bigEndian) { Data($0) }
    }
}
